### ISSUE TITLE
sqlite3 first time deploy doesn't work if ~/.sqliterc exists

### ISSUE THREAD

**Issue by kyzn** [sqitchers/sqitch#460]
Hi! Thanks for maintaining sqitch, it's a great library. I hope you find this issue (and proposed solution to it) helpful. Here goes.

I've recently tried to deploy a sqitch for the first time in a library (so no sqitch.db exists yet). It failed with an interesting error message:

```
Adding registry tables to db:sqlite:sqitch.db
-- Loading resources from /home/kivanc/.sqliterc
Use of uninitialized value in split at /home/kivanc/perl5/perlbrew/perls/perl-5.28.1/lib/site_perl/5.28.1/App/Sqitch/Engine/sqlite.pm line 117.
Use of uninitialized value $v[0] in numeric gt (>) at /home/kivanc/perl5/perlbrew/perls/perl-5.28.1/lib/site_perl/5.28.1/App/Sqitch/Engine/sqlite.pm line 120.
Use of uninitialized value $v[0] in numeric eq (==) at /home/kivanc/perl5/perlbrew/perls/perl-5.28.1/lib/site_perl/5.28.1/App/Sqitch/Engine/sqlite.pm line 120.
Sqitch requires SQLite 3.3.9 or later; sqlite3 is
```

Now, it looks like it can't get my sqlite3 version, which happens with running `sqlite3 -version` through `IPC::System::Simple`'s `capturex`. Also note the message about `Loading resources`.

I figured what happens is related to those two. If I don't have a `.sqliterc` file in my home directory, this is the output from `sqlite3 -version`:

```
0> use IPC::System::Simple qw/capturex/
1> capturex('sqlite3','-version');
$res[0] = '3.11.0 2016-02-15 17:29:24 3d862f207e3adc00f78066799ac5a8c282430a5f
'
```

And if I do have it, here's what happens:
```
2> capturex('sqlite3','-version');
-- Loading resources from /home/kivanc/.sqliterc
$res[1] = [
  '
',
  '3.11.0 2016-02-15 17:29:24 3d862f207e3adc00f78066799ac5a8c282430a5f
'
]
```

What happened here? There's now some additional output that went to STDOUT. And capturex returned an array with two items, first one being a `\n`.

The way `App::Sqitch::Engine::sqlite` checks for installed sqlite3 version is this: It calls `App::Sqitch`'s `probe`, which is `App::Sqitch`'s `capture` + a `chomp`. `capture` calls `capturex` (from above) and returns whatever it returns.

So where's the issue? It's in `sub probe`:
```
485 sub probe {
486     my ($ret) = shift->capture(@_);
487     chomp $ret if $ret;
488     return $ret;
489 }
```
`capturex` returns an array with its first element `\n`. `capture` passes that array to `probe` as is. And since we have a `($ret)` here, we are assigning that `\n` to `$ret`, which is just an empty string after `chomp`. And definitely that's not the sqlite3 version I have installed.

For solution.. I was going to suggest making sure `capture` returns a scalar. So instead of getting an array, we would get this:

```
3> my $scalar_ret_val = capturex('sqlite3','-version');
-- Loading resources from /home/kivanc/.sqliterc
$res[2] = '
3.11.0 2016-02-15 17:29:24 3d862f207e3adc00f78066799ac5a8c282430a5f
'
```

If you like the solution, great! I will shortly submit a pull request for it. If you think there should be some other approach for this problem, I'm all ears.

Thanks!

**Comment by theory** [comment]
Thanks. Seems like the behavior of `capture` and `probe` should not change from how they're currently documented. Better, I think, to call `capture` in a scalar context from sqlite.pm, something like:

``` diff
--- a/lib/App/Sqitch/Engine/sqlite.pm
+++ b/lib/App/Sqitch/Engine/sqlite.pm
@@ -115,7 +115,7 @@ has _sqlite3 => (
 
         # Make sure we can use this version of SQLite.
         my @v = split /[.]/ => (
-            split / / => $self->sqitch->probe( $self->client, '-version' )
+            split / / => scalar $self->sqitch->capture( $self->client, '-version' )
         )[0];
         hurl sqlite => __x(
             'Sqitch requires SQLite 3.3.9 or later; {client} is {version}',
```
That will take some fiddling with the tests, though, since they currently mock `probe` to return a version. Would have to instead mock `capture` somehow.

**Comment by kyzn** [comment]
Thanks for reply! Let me look into your patch and see if I can make it play nicely with tests. I'll cancel current PR and likely submit another one soon.

**Comment by kyzn** [comment]
I have just tried this on a different machine (with a `.sqliterc` file), and it seems to be working just fine. So the issue is probably even less widespread.

**Comment by theory** [comment]
Huh. What's up with your `~/.sqliterc` that it changes the output of `sqlite3 -version`?

**Comment by kyzn** [comment]
Could it be different sqlite3 versions? On one machine, I have 3.11.0 which comes with an extra newline:

```
kivanc@peppy:~$ sqlite3 -version
-- Loading resources from /home/kivanc/.sqliterc

3.11.0 2016-02-15 17:29:24 3d862f207e3adc00f78066799ac5a8c282430a5f
kivanc@peppy:~$ sqlite3 test.db
```

On other machine I have 3.16.2 with one less `\n`.

```
kivancyazan@penguin:~$ sqlite3 -version
-- Loading resources from /home/kivancyazan/.sqliterc
3.16.2 2017-01-06 16:32:41 a65a62893ca8319e89e48b8a38cf8a59c69a8209
kivancyazan@penguin:~$

**Comment by theory** [comment]
Okay, so the key is to collect the entire output and skip over empty lines, as I suggested above. Breaks tests, though. Anyone got some TUITs to update them?

**Comment by kyzn** [comment]
I'm trying to find a way to fix failing tests.. will spend some time today & update this issue.

**Comment by kyzn** [comment]
With suggested patch, `t/sqlite.t` gets confused about version:

    Sqitch requires SQLite 3.3.9 or later; sqlite3 is 2

This happens on both my machines. I think 2 is coming from number of elements in array.

To be more specific, at

    64 is_deeply [$sqlite->sqlite3], [$sqlite->client, @std_opts, $sqlite->uri->dbname],
    65     'sqlite3 command should have the proper opts';

It does call `capture` and collects correct version. However, later it starts thinking version is `2`. Looking into `sqlite.t` now, but would appreciate any clues.

**Comment by kyzn** [comment]
I think I got it, patch incoming soon.

**Comment by theory** [comment]
Fixed thanks to @kyzn in d630dbe.
